assembly/src/tokens/mod.rs

```diff
@@ -233,79 +231,99 @@ impl<'a> fmt::Display for Token<'a> {
+        }
     }
 
-    Ok(label.to_string())
+    Ok(AbsolutePath::new_unchecked(path.to_string()))
 }
 
 /// Procedure locals must be a 16-bit integer.
 fn validate_proc_locals(locals: &str, token: &Token) -> Result<u16, ParsingError> {
     match locals.parse::<u64>() {
         Ok(num_locals) => {
             if num_locals > u16::MAX as u64 {
-                return Err(ParsingError::invalid_proc_locals(token, locals));
+                return Err(ParsingError::too_many_proc_locals(
+                    token,
+                    num_locals,
+                    u16::MAX as u64,
+                ));
             }
             Ok(num_locals as u16)
         }
         Err(_) => Err(ParsingError::invalid_proc_locals(token, locals)),
     }
 }
 
-/// A module import path must comply with the following rules:
-/// - It must start with an ascii letter.
-/// - It can contain only ascii letters, numbers, underscores, or colons.
-fn validate_import_path(path: &str, token: &Token) -> Result<String, ParsingError> {
-    // a path must start with a letter
-    if path.is_empty() || !path.chars().next().unwrap().is_ascii_alphabetic() {
-        return Err(ParsingError::invalid_module_path(token, path));
+/// A label of an invoked procedure must comply with the following rules:
+/// - It can contain a single procedure name. In this case, the label must comply with procedure
+///   name rules.
+/// - It can contain module name followed by procedure name (e.g., "module::procedure"). In this
+///   case both module and procedure name must comply with relevant name rules.
+///
+/// All other combinations will result in an error.
+///
+/// TODO: validation should happen at the path and procedure name levels rather than here.
+fn validate_proc_invocation_label<'a>(
+    label: &'a str,
+    token: &'a Token,
+) -> Result<(&'a str, Option<&'a str>), ParsingError> {
+    // a label must start with a letter
+    if label.is_empty() || !label.chars().next().unwrap().is_ascii_alphabetic() {
+        return Err(ParsingError::invalid_proc_invocation(token, label));
     }
 
-    // a path can contain only letters, numbers, underscores, or colons
-    if !path
-        .chars()
-        .all(|c| c.is_ascii_alphanumeric() || c == '_' || c == ':')
-    {
-        return Err(ParsingError::invalid_module_path(token, path));
+    let mut parts = label.split(MODULE_PATH_DELIM);
+    let (proc_name, module_name) = match (parts.next(), parts.next()) {
+        (None, _) => return Err(ParsingError::invalid_proc_invocation(token, label)),
+        (Some(proc_name), None) => {
+            if !is_valid_label(proc_name) {
+                return Err(ParsingError::invalid_proc_invocation(token, label));
+            }
+            (proc_name, None)
+        }
+        (Some(module_name), Some(proc_name)) => {
+            if !is_valid_label(proc_name) || !is_valid_label(module_name) || parts.next().is_some()
```

Comment:
This part confuses me a bit. Why don't we allow, given for example `std::foo::bar::baz`, `use.std::foo` and `call.foo::bar::baz`? Is it implementation convenience or a ruler we actually want to enforce? It is feasible to compute the procedure id for both so maybe we shouldn't enforce this restriction.

Reply:
I think it is mostly implementation complexity. I think it might be good to allow absolute paths too. So, maybe we should create an issue for this.

Reply:
https://github.com/maticnetwork/miden/issues/587